sdk/internal/temporal/resource.go

```diff
@@ -0,0 +1,120 @@
+	// If the resource is expiring within this time window, update it eagerly.
+	// This allows other threads/goroutines to keep running by using the not-yet-expired
+	// resource value while one thread/goroutine updates the resource.
+	const window = 5 * time.Minute   // This example updates the resource 5 minutes prior to expiration
+	const backoff = 30 * time.Second // Minimum wait time between eager update attempts
+
+	now, acquire, expired, resource := time.Now(), false, false, er.resource
+	// acquire exclusive lock
+	er.cond.L.Lock()
+	for {
+		expired = er.expiration.IsZero() || er.expiration.Before(now)
+		if expired {
+			// The resource was never acquired or has expired
+			if !er.acquiring {
+				// If another thread/goroutine is not acquiring/updating the resource, this thread/goroutine will do it
+				er.acquiring, acquire = true, true
+				break
+			}
+			// Getting here means that this thread/goroutine will wait for the updated resource
+		} else if er.expiration.Add(-window).Before(now) {
+			// The resource is valid but is expiring within the time window
+			if !er.acquiring && er.lastAttempt.Add(backoff).Before(now) {
+				// If another thread/goroutine is not acquiring/renewing the resource, and none has attempted
+				// to do so within the last 30 seconds, this thread/goroutine will do it
+				er.acquiring, acquire = true, true
+				break
+			}
+			// This thread/goroutine will use the existing resource value while another updates it
+			resource = er.resource
+			break
+		} else {
+			// The resource is not close to expiring, this thread/goroutine should use its current value
+			resource = er.resource
+			break
+		}
+		// If we get here, wait for the new resource value to be acquired/updated
+		er.cond.Wait()
+	}
+	er.cond.L.Unlock() // Release the lock so no threads/goroutines are blocked
+
+	var err error
+	if acquire {
+		// This thread/goroutine has been selected to acquire/update the resource
+		var expiration time.Time
+		var newValue TResource
+		er.lastAttempt = now
+		newValue, expiration, err = er.acquireResource(state)
+
+		// Atomically, update the shared resource's new value & expiration.
+		er.cond.L.Lock()
+		if err == nil {
+			// Update resource & expiration, return the new value
+			resource = newValue
+			er.resource, er.expiration = resource, expiration
+		} else if !expired {
+			// An eager update failed. Discard the error and return the current--still valid--resource value
+			err = nil
+		}
+		er.acquiring = false // Indicate that no thread/goroutine is currently acquiring the resource
+
+		// Wake up any waiting threads/goroutines since there is a resource they can ALL use
+		er.cond.L.Unlock()
+		er.cond.Broadcast()
+	}
+	return resource, err // Return the resource this thread/goroutine can use
+}
+
+// Expire marks the resource as expired, ensuring it's refreshed on the next call to GetResource().
+func (er *Resource[TResource, TState]) Expire() {
```

Comment:
This is a new method.  It's used by the key vault challenge policy.

Reply:
Why does Key Vault's policy need to force refresh a token?

Reply:
Good question.  Honestly, I'm not sure as I wasn't involved in this implementation.

Per the [comments](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/keyvault/internal/challenge_policy.go#L88-L91).

```go
// If it fails and has a 401, try it with a new token
```

@schaabs do you have any context on why we might want to explicitly expire a token credential in this scenario?

Reply:
The policy discards its cached token when it sees a 401 because the error means the token is invalid despite not being expired. What I'm getting at is whether the policy needs `Expire()` to do this. I would think it can just create a new `ExpiringResource`, which makes sense conceptually because the policy may need to acquire a token in a new tenant, or with new claims.

Reply:
I see your point.  Presumably using `Expire()` would be a bit more GC-friendly in this case compared to creating a new `ExpiringResource`.  However, it's probably infrequent enough that it doesn't really matter (well, I'm guessing here but I suspect this is why it was implemented in this way).

Reply:
It should be a rare event and even if not, I expect the perf impact due to collection is trivial compared to the impact of reauthenticating and resending the service request.

Reply:
At this point I'm inclined to leave it given that key vault has tested with this implementation.  It's also not available to the public.  Do you feel particularly strong about removing it?

Reply:
I think it's okay for now, being internal, but Key Vault's policy needs another look after #14932